zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java

```diff
@@ -325,18 +326,39 @@ private void validateParameters(final DirectBuffer type) {
   }
 
   @Override
-  public void forEachTimedOutEntry(
-      final long upperBound, final BiPredicate<Long, JobRecord> callback) {
+  public DeadlineIndex forEachTimedOutEntry(
+      final long executionTimestamp,
+      final DeadlineIndex startAt,
+      final BiPredicate<Long, JobRecord> callback) {
+
+    final DbCompositeKey<DbLong, DbForeignKey<DbLong>> startAtKey;
+    if (startAt != null) {
+      deadlineKey.wrapLong(startAt.deadline());
+      jobKey.wrapLong(startAt.key());
+      startAtKey = deadlineJobKey;
+    } else {
+      startAtKey = null;
+    }
+
+    final var nextIndex = new AtomicReference<DeadlineIndex>();
     deadlinesColumnFamily.whileTrue(
+        startAtKey,
         (key, value) -> {
-          final long deadline = key.first().getValue();
-          final boolean isDue = deadline < upperBound;
-          if (isDue) {
-            final long jobKey1 = key.second().inner().getValue();
-            return visitJob(jobKey1, callback);
+          final var deadline = key.first().getValue();
+          final var isDue = deadline < executionTimestamp;
+          if (!isDue) {
+            return false;
+          }
+          final var jobKey = key.second().inner().getValue();
+          if (!visitJob(jobKey, callback)) {
+            nextIndex.set(
+                new DeadlineIndex(key.first().getValue(), key.second().inner().getValue()));
```

Comment:
That's not correct @mustafadagher. Consider the interface of this method here. Is the visitor called for this index yes/no? It is, but the visitor returned false, indicating that it stops here. The caller then receives the last visited index back so it can use it to start there again next time. This is the last visited index and the one to start with next time.

But it's not the next index. Because that would mean that the visitor did not yet see this entry. However, it did already see it and it responded false to it.

Reply:
okay, I'm kinda convinced haha. will change it :)